### ISSUE TITLE
local_port_forwarding

### ISSUE THREAD

**Issue by wmegchel** [ronf/asyncssh#410]
Hi! 

I succeeded in forwarding a port from a server "XX" by running:

`async with asyncssh.connect('XX') as conn:`
`listener = await conn.forward_local_port('', 9000, 'localhost', 9000)`
`await listener.wait_closed()`

Now, I would like to forward a port from a server  (say W1)  connected to XX (a HPC worker node). If I do it on the command line:

`ssh -L localhost:9000:W1:9000 XX `
works fine.

So I tried
`listener = await conn.forward_local_port('', 9000, 'W1', 9000)` or 
`listener = await conn.forward_local_port('localhost', 9000, 'W1', 9000)`

but that does not work. Can I make it work using asyncssh, and if so how?

Thanks!

**Comment by ronf** [comment]
Assuming that the SSH server "XX" can resolve the host "W1", that the "W1" host has a listener on port 9000, and that port forwarding is allowed on "XX" to that destination, I see no reason why this shouldn't work. What sort of failure are you seeing?

Also, just to be clear, this should be setting up a listener on the host where you are running AsyncSSH which listens on the local port 9000 and forwards the data over the SSH connection to the listener on port 9000 of host "W1".

**Comment by wmegchel** [comment]
Thanks for your quick reply!

_Assuming that the SSH server "XX" can resolve the host "W1", that the "W1" host has a listener on port 9000, and that port forwarding is allowed on "XX" to that destination, I see no reason why this shouldn't work._ 

Good to hear. I assumed that this should work as well. What I am doing is forwarding an Rstudio server running on port 9000 from a compute node (W1) on a HPC (login node=XX) towards my own computer. If I use "ssh -L localhost:9000:W1:9000 XX"  from a BASH shell, I can start Rstudio from `localhost:9000` in my browser, so that seems to be OK.

_What sort of failure are you seeing?_
With `listener = await conn.forward_local_port('', 9000, 'W1', 9000)`, shows no error. The python script just terminates when I open my browser with at "localhost:9000". I know asyncssh can log detailed (debug) info, but I don't know how...

The same script works fine if I connect to a server directly and use:
`listener = await conn.forward_local_port('', 9000, localhost, 9000)`

Opening `localhost:9000` then opens Rstudio as intented on my PC.

In short, I think I am making a mistake with the forward command, but I don't know what is wrong.

**Comment by ronf** [comment]
If you have an "await listener.wait_closed()" but you're still seeing the script exit, it's probably raising an exception of some kind. You can try adding the following to enable logging:

```python
import logging
logging.basicConfig(level='DEBUG')
asyncssh.set_debug_level(2)
```

Can you show a more complete example, including the call to asyncio to run the event loop?

Also, make sure the calls that use "conn" are all indented inside the "async with" block that assigns it. Once you leave that block, the connection will be closed.

**Comment by wmegchel** [comment]
Hi Ron, 

This is the full script:

`import asyncssh, asyncio, sys, os
from contextlib import suppress
from sshtunnel import SSHTunnelForwarder
import logging

host = 'cart'
worker = "tcn995"
passphrase = "XXXXXXX"
port = 9000

logging.basicConfig(level='DEBUG')
asyncssh.set_debug_level(2)

# forward port to worker
async def run_client():
    async with asyncssh.connect(host, passphrase=passphrase) as conn:
        listener = await conn.forward_local_port('', port, worker, port)
        await listener.wait_closed()

try:
    asyncio.get_event_loop().run_until_complete(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit('SSH connection failed: ' + str(exc))
`

and indeed I get a uncaught exception now that I've set debug level=2. This is the full debug info:

---
forwarding port 9000 to localhost
DEBUG:asyncio:Using selector: EpollSelector
INFO:asyncssh:[conn=0] Connected to SSH server at cartesius.surfsara.nl, port 22
INFO:asyncssh:[conn=0]   Proxy command: ssh -xaqWcartesius.surfsara.nl:22 hpc
DEBUG:asyncssh:[conn=0] Sending version SSH-2.0-AsyncSSH_2.7.0
DEBUG:asyncssh:[conn=0] Received version SSH-2.0-OpenSSH_7.4
DEBUG:asyncssh:[conn=0] Requesting key exchange
DEBUG:asyncssh:[conn=0]   Key exchange algs: curve25519-sha256,curve25519-sha256@libssh.org,curve448-sha512,ecdh-sha2-nistp521,ecdh-sha2-nistp384,ecdh-sha2-nistp256,ecdh-sha2-1.3.132.0.10,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha256,diffie-hellman-group15-sha512,diffie-hellman-group16-sha512,diffie-hellman-group17-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256@ssh.com,diffie-hellman-group14-sha1,rsa2048-sha256
DEBUG:asyncssh:[conn=0]   Host key algs: ecdsa-sha2-nistp256
DEBUG:asyncssh:[conn=0]   Encryption algs: chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
DEBUG:asyncssh:[conn=0]   MAC algs: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1,hmac-sha256-2@ssh.com,hmac-sha224@ssh.com,hmac-sha256@ssh.com,hmac-sha384@ssh.com,hmac-sha512@ssh.com
DEBUG:asyncssh:[conn=0]   Compression algs: zlib@openssh.com,none
DEBUG:asyncssh:[conn=0] Received key exchange request
DEBUG:asyncssh:[conn=0]   Key exchange algs: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
DEBUG:asyncssh:[conn=0]   Host key algs: ssh-rsa,rsa-sha2-512,rsa-sha2-256,ssh-dss,ecdsa-sha2-nistp256,ssh-ed25519
DEBUG:asyncssh:[conn=0]   Client to server:
DEBUG:asyncssh:[conn=0]     Encryption algs: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,aes192-cbc,aes256-cbc,blowfish-cbc,cast128-cbc,3des-cbc
DEBUG:asyncssh:[conn=0]     MAC algs: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG:asyncssh:[conn=0]     Compression algs: none,zlib@openssh.com
DEBUG:asyncssh:[conn=0]   Server to client:
DEBUG:asyncssh:[conn=0]     Encryption algs: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,aes192-cbc,aes256-cbc,blowfish-cbc,cast128-cbc,3des-cbc
DEBUG:asyncssh:[conn=0]     MAC algs: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG:asyncssh:[conn=0]     Compression algs: none,zlib@openssh.com
DEBUG:asyncssh:[conn=0] Beginning key exchange
DEBUG:asyncssh:[conn=0]   Key exchange alg: curve25519-sha256
DEBUG:asyncssh:[conn=0]   Client to server:
DEBUG:asyncssh:[conn=0]     Encryption alg: chacha20-poly1305@openssh.com
DEBUG:asyncssh:[conn=0]     MAC alg: chacha20-poly1305@openssh.com
DEBUG:asyncssh:[conn=0]     Compression alg: zlib@openssh.com
DEBUG:asyncssh:[conn=0]   Server to client:
DEBUG:asyncssh:[conn=0]     Encryption alg: chacha20-poly1305@openssh.com
DEBUG:asyncssh:[conn=0]     MAC alg: chacha20-poly1305@openssh.com
DEBUG:asyncssh:[conn=0]     Compression alg: zlib@openssh.com
DEBUG:asyncssh:[conn=0] Requesting service ssh-userauth
DEBUG:asyncssh:[conn=0] Completed key exchange
DEBUG:asyncssh:[conn=0] Received extension info
DEBUG:asyncssh:[conn=0]   server-sig-algs: rsa-sha2-256,rsa-sha2-512
DEBUG:asyncssh:[conn=0] Request for service ssh-userauth accepted
INFO:asyncssh:[conn=0] Beginning auth for user woutm
DEBUG:asyncssh:[conn=0] Remaining auth methods: publickey,gssapi-keyex,gssapi-with-mic,password,keyboard-interactive
DEBUG:asyncssh:[conn=0] Preferred auth methods: gssapi-keyex,gssapi-with-mic,hostbased,publickey,keyboard-interactive,password
DEBUG:asyncssh:[conn=0] Trying public key auth with rsa-sha2-256 key
DEBUG:asyncssh:[conn=0] Signing request with rsa-sha2-256 key
INFO:asyncssh:[conn=0] Auth for user woutm succeeded
INFO:asyncssh:[conn=0] Creating local TCP forwarder from port 9000 to tcn995, port 9000
DEBUG:asyncssh:[conn=0] Received unknown global request: hostkeys-00@openssh.com
INFO:asyncssh:[conn=0] Opening direct TCP connection to tcn995, port 9000
INFO:asyncssh:[conn=0]   Client address: ::1, port 33500
DEBUG:asyncssh:[conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
DEBUG:asyncssh:[conn=0, chan=0]   Initial recv window 2097152, packet size 32768
INFO:asyncssh:[conn=0] Opening direct TCP connection to tcn995, port 9000
INFO:asyncssh:[conn=0]   Client address: ::1, port 33502
DEBUG:asyncssh:[conn=0, chan=1] Set write buffer limits: low-water=16384, high-water=65536
DEBUG:asyncssh:[conn=0, chan=1]   Initial recv window 2097152, packet size 32768
DEBUG:asyncssh:[conn=0, chan=0]   Initial send window 2097152, packet size 32768
DEBUG:asyncssh:[conn=0] Uncaught exception
Traceback (most recent call last):
  File "/home/wmegchel/software/miniconda3/lib/python3.8/site-packages/asyncssh/connection.py", line 829, in _reap_task
    task.result()
  File "/home/wmegchel/software/miniconda3/lib/python3.8/site-packages/asyncssh/forward.py", line 152, in _forward
    await self._coro(session_factory, *args)
  File "/home/wmegchel/software/miniconda3/lib/python3.8/site-packages/asyncssh/connection.py", line 2579, in tunnel_connection
    return (await self.create_connection(session_factory,
  File "/home/wmegchel/software/miniconda3/lib/python3.8/site-packages/asyncssh/connection.py", line 3705, in create_connection
    session = await chan.connect(session_factory, remote_host, remote_port,
  File "/home/wmegchel/software/miniconda3/lib/python3.8/site-packages/asyncssh/channel.py", line 1902, in connect
    return (await self._open_tcp(session_factory, b'direct-tcpip',
  File "/home/wmegchel/software/miniconda3/lib/python3.8/site-packages/asyncssh/channel.py", line 1894, in _open_tcp
    return (await super()._open_forward(session_factory, chantype,
  File "/home/wmegchel/software/miniconda3/lib/python3.8/site-packages/asyncssh/channel.py", line 1876, in _open_forward
    self._session.connection_made(self)
  File "/home/wmegchel/software/miniconda3/lib/python3.8/site-packages/asyncssh/forward.py", line 79, in connection_made
    if sock.family in {socket.AF_INET, socket.AF_INET6}:
AttributeError: 'NoneType' object has no attribute 'family'
INFO:asyncssh:[conn=0] Connection failure: 'NoneType' object has no attribute 'family'
INFO:asyncssh:[conn=0, chan=0] Closing channel due to connection close
INFO:asyncssh:[conn=0, chan=0] Closing channel
INFO:asyncssh:[conn=0, chan=0] Channel closed: 'NoneType' object has no attribute 'family'
INFO:asyncssh:[conn=0, chan=1] Closing channel due to connection close
INFO:asyncssh:[conn=0, chan=1] Channel closed: 'NoneType' object has no attribute 'family'
INFO:asyncssh:[conn=0] Closing connection
INFO:asyncssh:[conn=0] Sending disconnect: Disconnected by application (11)

I hope you know what is wrong. Thanks again! 

Wout

**Comment by ronf** [comment]
The test script looks fine here, and up until the point of the uncaught exception, I don't really see anything wrong in the log output, but I did notice that the application code seems to be opening two tunneled connections at once (one from port 33500 and another from port 33502). I wonder maybe one of those two connections is being opened and immediately closed, before AsyncSSH has had a chance to finish connecting upstream, and perhaps there's a race condition here.

The problem seems to be that the call to `transport.get_extra_info('socket')` at line 78 in forward.py is returning `None`. The transport at that point should be an SSHTCPChannel object, and any queries for extra info would generally be passed on to the SSH connection object, returning the socket that the SSH connection itself is using. In the process of cleaning up a channel, the reference to the connection object from the channel object could be set to `None`, so AsyncSSH may need a guard against that.

This is only a guess, but you might want to try changing the connection_made() call in SSHForward in forward.py from:

```python
        sock = transport.get_extra_info('socket')
        if sock.family in {socket.AF_INET, socket.AF_INET6}:
            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
```

to:

```python
        sock = transport.get_extra_info('socket')
        if sock and sock.family in {socket.AF_INET, socket.AF_INET6}:
            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
```

Any idea why your test application would be opening two ports at once here? Do you think it is possible that one of those two ports is being closed immediately?

**Comment by wmegchel** [comment]
Hi Ron, 

Adapted the code like you suggested and that solved the issue! I have no clue why the other ports (33500 and 33502) are used. I also don't see how these are related to my request to forward on port 9000... 

In any case, I am very happy that it works now. Maybe you can implement this small change in your next release. Thanks again for your rapid response.

**Comment by ronf** [comment]
This is now available as commit c023f0a in the "develop" branch and will be included in the next release.

**Comment by wmegchel** [comment]
Thanks Ron! I will close the issue.
